### PR TITLE
Update minimum rust version to 1.72

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/8743

**NOTE**: Update to update to MSRV 1.72 (was 1.71)

## Rationale for this change

1. the new version of `env_logger` requires rust 1.71: https://github.com/apache/arrow-datafusion/pull/8944
2. Current MSRV of 1.70 which was released more than 6 months ago: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1700-2023-06-01 (ages in Rust time)

MSRV 1.72 is still 5 months ago:  in 2023-08-24   https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1720-2023-08-24

Most recent rust is 1.75: released end of Dec 2023
https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1750-2023-12-28

## What changes are included in this PR?
Update Minimum rust version to 1.72

## Are these changes tested?
Covered by existing tests

### CODE REVIEW

File: datafusion/core/Cargo.toml
Comment (Jefffrey) [inline datafusion/core/Cargo.toml:30]: Could we take this opportunity to set this as same as workspace? e.g.

https://github.com/apache/arrow-datafusion/blob/7a5f2054305fd92852b589473afbc9bb034379d7/datafusion/physical-plan/Cargo.toml#L29

Unless its preferable to keep these separate? (even though they currently are the same)

Comment (alamb) [reply datafusion/core/Cargo.toml:30]: That is an excellent idea.  I vaguely (but perhaps incorrectly) remember something related to releasing was tricky with MSRV -- @andygrove  do you remember? Would it be ok to switch the sub crates to use the workspace version?

Comment (alamb) [reply datafusion/core/Cargo.toml]: Done in 9280b7a166814bc6f3660c7471fb9858ac8bde3a

**Comment by ozankabak** [comment]
Should we upgrade to 1.75 to avoid two upgrades? The latest version of one of our dependencies, `ahash`, requires 1.75 and we will hit problems sooner or later

**Comment by alamb** [comment]
> Should we upgrade to 1.75 to avoid two upgrades? The latest version of one of our dependencies, `ahash`, requires 1.75 and we will hit problems sooner or later

Going up to the latest stable would be good from my perspective. 

I think we should decide what our policy is and document it. @comphead 's recent change https://github.com/apache/arrow-datafusion/pull/9071 documents what the test is but not what the policy is (aka when we will update the value). 

@Ted-Jiang  also filed https://github.com/apache/arrow-datafusion/issues/8743

Here is my proposal -- I'll a "MSRV policy" ticket with proposal, and we can have a larger community discussion around that.

**Comment by alamb** [comment]
Ok, after some more consideration, I propose we change this PR to update to 1.72 to close @Ted-Jiang's https://github.com/apache/arrow-datafusion/issues/8743 and hopefully unblock ahash updates, and then we have the larger discussion about MSRV policy with the wider community

File: datafusion/core/Cargo.toml
Comment (alamb) [reply datafusion/core/Cargo.toml]: Ok, I found out again. The reason the version is replicated is that `cargo msrv verify` doesn't support workspace versions

Thus if you do `cd datafusion/core && cargo msrv verify`, it fails like 

```
Fetching index
Unable to find key 'package.rust-version' (or 'package.metadata.msrv') in '/Users/andrewlamb/Software/arrow-datafusion/Cargo.toml'
```

I have fixed that

File: datafusion/expr/src/expr.rs
Comment (alamb) [inline datafusion/expr/src/expr.rs:856]: Clippy failed on this for me with the updated MSRV so I did what it told me

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb let me know if we need a review for MSRV policy or dscussion

**Comment by alamb** [comment]
> lgtm thanks @alamb let me know if we need a review for MSRV policy or dscussion

Thanks @comphead  -- let's have the MSRV policy discussion on https://github.com/apache/arrow-datafusion/issues/9082

File: datafusion/core/Cargo.toml
Comment (Jefffrey) [reply datafusion/core/Cargo.toml]: Tracking issue for that in cargo msrv: https://github.com/foresterre/cargo-msrv/issues/590

**Review by Jefffrey — APPROVED** [review]
:+1:

File: datafusion/core/Cargo.toml
Comment (alamb) [inline datafusion/core/Cargo.toml:32]:
```suggestion
# https://github.com/foresterre/cargo-msrv/issues/590
rust-version = "1.72"
```

File: datafusion/core/Cargo.toml
Comment (alamb) [reply datafusion/core/Cargo.toml]: Thanks @Jefffrey  -- I have added an issue ref as a comment so we don't forget it in the future

**Comment by Omega359** [comment]
@alamb You missed the cli Dockerfile unfortunately - it's still 1.70 and won't build without bumping up the version.

**Comment by alamb** [comment]
> @alamb You missed the cli Dockerfile unfortunately - it's still 1.70 and won't build without bumping up the version.

Will fix

**Pull request merged**
